samples/nrf9160/cloud_simple/src/main.c

```diff
@@ -0,0 +1,249 @@
+}
+
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf)
+{
+	ARG_UNUSED(esf);
+
+	LOG_PANIC();
+	printk("Running main.c error handler");
+	error_handler(ERROR_SYSTEM_FAULT, reason);
+	CODE_UNREACHABLE;
+}
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	error_handler(ERROR_BSD_RECOVERABLE, (int)err);
+}
+
+/**@brief Routine that publishes a custom message to cloud. */
+static void cloud_update_routine_work_fn(struct k_work *work)
+{
+	int err;
+
+	printk("Publishing message: %s\n", CONFIG_CLOUD_MESSAGE);
+
+	struct cloud_msg msg = {
+		.qos = CLOUD_QOS_AT_MOST_ONCE,
+		.endpoint.type = CLOUD_EP_TOPIC_MSG,
+		.buf = CONFIG_CLOUD_MESSAGE,
+		.len = sizeof(CONFIG_CLOUD_MESSAGE)
+	};
+
+	err = cloud_send(cloud_backend, &msg);
+	if (err) {
+		printk("cloud_send failed, error: %d\n", err);
+	}
+
+	k_delayed_work_submit(
+			&cloud_update_routine_work,
+			K_SECONDS(CONFIG_CLOUD_MESSAGE_PUBLICATION_INTERVAL));
+}
+
+void cloud_event_handler(const struct cloud_backend *const backend,
+			 const struct cloud_event *const evt,
+			 void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	switch (evt->type) {
+	case CLOUD_EVT_CONNECTED:
+		printk("CLOUD_EVT_CONNECTED\n");
+		break;
+	case CLOUD_EVT_READY:
+		printk("CLOUD_EVT_READY\n");
+		k_delayed_work_submit(&cloud_update_routine_work, K_NO_WAIT);
+		break;
+	case CLOUD_EVT_DISCONNECTED:
+		printk("CLOUD_EVT_DISCONNECTED\n");
+		break;
+	case CLOUD_EVT_ERROR:
+		printk("CLOUD_EVT_ERROR\n");
+		break;
+	case CLOUD_EVT_DATA_SENT:
+		printk("CLOUD_EVT_DATA_SENT\n");
+		break;
+	case CLOUD_EVT_DATA_RECEIVED:
+		printk("CLOUD_EVT_DATA_RECEIVED\n");
+		printk("Data received from cloud: %s\n", evt->data.msg.buf);
+		break;
+	case CLOUD_EVT_PAIR_REQUEST:
+		printk("CLOUD_EVT_PAIR_REQUEST\n");
+		break;
+	case CLOUD_EVT_PAIR_DONE:
+		printk("CLOUD_EVT_PAIR_DONE\n");
+		break;
+	case CLOUD_EVT_FOTA_DONE:
+		printk("CLOUD_EVT_FOTA_DONE\n");
+		break;
+	default:
+		printk("Unknown cloud event type: %d\n", evt->type);
+		break;
+	}
+}
+
+/**@brief Initialize delayed work items. */
+static void work_init(void)
+{
+	k_delayed_work_init(&cloud_update_routine_work,
+			    cloud_update_routine_work_fn);
+}
+
+/**@brief Configures modem to provide LTE link. Blocks until link is
+ * successfully established.
+ */
+static void modem_configure(void)
+{
+#if defined(CONFIG_BSD_LIBRARY)
+	if (IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT)) {
+		/* Do nothing, modem is already turned on
+		 * and connected.
+		 */
+	} else {
+		int err;
+
+		printk("Connecting to LTE network. ");
+		printk("This may take several minutes.\n");
+
+		err = lte_lc_init_and_connect();
+		if (err) {
+			printk("LTE link could not be established.\n");
+			error_handler(ERROR_LTE_LC, err);
+		}
+
+		printk("Connected to LTE network\n");
+	}
+#endif
+}
+
+void main(void)
+{
+	int err;
+
+	printk("Cloud simple has started\n");
+
+	cloud_backend = cloud_get_binding("NRF_CLOUD");
```

Comment:
We should create a KConfig variable that would allow to switch between cloud backends. It's counter-intuitive to have some of the sample tweaks done at the Kconfig level (like the message content) and some from the source-code level.

Reply:
Agreed